indexes: Stop using node internal types

ryanofsky opened the pull request:
Start transitioning index code away from using internal node types like `CBlockIndex` and `CChain` so index code is less coupled to node code and index code will later be able to stop locking cs_main and sync without having to deal with validationinterface race conditions, and so new indexes are easier to write and can run as plugins or separate processes.

This PR contains the first 7 commits from https://github.com/bitcoin/bitcoin/pull/24230#issuecomment-1165625977 which have been split off for easier review. Previous review comments can be found in #24230

ryanofsky committed:
interfaces, refactor: Add more block information to block connected notifications

Add new interfaces::BlockInfo struct to be able to pass extra block
information (file and undo information) to indexes which they are
updated to use high level interfaces::Chain notifications.

This commit does not change behavior in any way.

ryanofsky committed:
indexes, refactor: Pass Chain interface instead of CChainState class to indexes

Passing abstract Chain interface will let indexes run in separate
processes.

This commit does not change behavior in any way.

ryanofsky committed:
indexes, refactor: Remove CBlockIndex* uses in coinstatsindex LookUpOne function

This commit does not change behavior in any way.

ryanofsky committed:
indexes, refactor: Remove CBlockIndex* uses in index Init methods

Replace overriden index Init() methods that use the best block
CBlockIndex* pointer with pure CustomInit() callbacks that are passed
the block hash and height.

This gets rid of more CBlockIndex* pointer uses so indexes can work
outside the bitcoin-node process. It also simplifies the initialization
call sequence so index implementations are not responsible for
initializing the base class.

There is a slight change in behavior here since now the best block
pointer is loaded and checked before the custom index init functions are
called instead of while they are called.

ryanofsky committed:
indexes, refactor: Remove CBlockIndex* uses in index WriteBlock methods

Replace WriteBlock method with CustomAppend and pass BlockInfo struct
instead of CBlockIndex* pointer

This commit does not change behavior in any way.

ryanofsky committed:
indexes, refactor: Remove CBlockIndex* uses in index Rewind methods

Replace Rewind method with CustomRewind and pass block hashes and
heights instead of CBlockIndex* pointers

This commit does not change behavior in any way.

ryanofsky committed:
indexes, refactor: Remove CChainState use in index CommitInternal method

Replace CommitInternal method with CustomCommit and use interfaces::Chain
instead of CChainState to generate block locator.

This commit does not change behavior in any way, except in the
(m_best_block_index == nullptr) case, which was added recently in
https://github.com/bitcoin/bitcoin/pull/24117 as part of an ongoing attempt to
prevent index corruption if bitcoind is interrupted during startup. New
behavior in that case should be slightly better than the old behavior (skipping
the entire custom+base commit now vs only skipping the base commit previously)
and this might avoid more cases of corruption.

MarcoFalke commented on src/node/chain.cpp:
```diff
@@ -0,0 +1,23 @@
+        info.prev_hash = index->pprev ? index->pprev->phashBlock : nullptr;
+        info.height = index->nHeight;
+        LOCK(::cs_main);
+        info.file_number = index->nFile;
```
It seems that this will only copy/alias the internal memory to a different type. I wonder if this is actually useful, because it disables lock annotations for the memory where they may actually still be needed? Also, I wonder how useful an interface is that returns a data structure that points into memory that is owned by a different process (I am assuming each blockindex runs in a different multiprocess)

fjahr commented:
Linter warns of a circular dependency

```
A new circular dependency in the form of "index/base -> node/context -> net_processing -> index/blockfilterindex -> index/base" appears to have been introduced.
```

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/25494#issuecomment-1172976870

> Linter warns of a circular dependency

Thanks! Should be fixed

---

Updated 3c82fa533a31939b867c774fe2c05f7a2e1d1a0c -> ddbbdcf4112c3a84e3287e6bf701e0eb592b24c3 ([`pr/ind.1`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.1) -> [`pr/ind.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ind.1..pr/ind.2)) to fix `lint-circular-dependencies.py` error

mzumsande reviewed (commented):
ACK ddbbdcf4112c3a84e3287e6bf701e0eb592b24c3

I reviewed these commits multiple times as part of #24230, and think they are good.
I also did some manual testing for this PR on regtest and signet and didn't see any differences in behavior.

While it's a bit unfortunate that this introduces a circular dependency, this is transitory and will get removed with the last commits of #24230 after all internal uses of `CChainState` are removed.

nit: clang-format suggests a few minor changes (spacing) over these commits

fjahr reviewed (commented):
tACK ddbbdcf4112c3a84e3287e6bf701e0eb592b24c3

Reviewed code and ran it, syncing `txindex` from intermittently synced state to fully synced and fully synced `coinstatsindex` from scratch.

MarcoFalke commented on src/Makefile.am:
```diff
@@ -878,6 +878,7 @@ libbitcoinkernel_la_SOURCES = \
   key.cpp \
   logging.cpp \
   node/blockstorage.cpp \
+  node/chain.cpp \
```
88cad3cb604058d8f26790eeb37abe2dabc0c033: Could explain in the commit message that this is only temporary? If not, move it to the kernel folder?

MarcoFalke commented on src/index/blockfilterindex.cpp:
```diff
@@ -209,22 +210,25 @@ size_t BlockFilterIndex::WriteFilterToDisk(FlatFilePos& pos, const BlockFilter&
         }
 
-        uint256 expected_block_hash = pindex->pprev->GetBlockHash();
+        uint256 expected_block_hash = *block.prev_hash;
```
88cad3cb604058d8f26790eeb37abe2dabc0c033: I wonder where UB should be avoided with asserts and where not

MarcoFalke commented on src/index/blockfilterindex.cpp:
```diff
@@ -233,18 +237,18 @@ bool BlockFilterIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex
     }
 
-    BlockFilter filter(m_filter_type, block, block_undo);
+    BlockFilter filter(m_filter_type, *block.data, block_undo);
```
Same

MarcoFalke commented on src/index/coinstatsindex.cpp:
```diff
@@ -110,24 +110,27 @@ CoinStatsIndex::CoinStatsIndex(std::unique_ptr<interfaces::Chain> chain, size_t
         }
 
-        uint256 expected_block_hash{pindex->pprev->GetBlockHash()};
+        uint256 expected_block_hash{*block.prev_hash};
```
Same

MarcoFalke commented on src/index/base.cpp:
```diff
@@ -206,22 +215,19 @@ void BaseIndex::ThreadSync()
     // Don't commit anything if we haven't indexed any block yet
     // (this could happen if init is interrupted).
     if (m_best_block_index == nullptr) {
         return false;
```
ddbbdcf4112c3a84e3287e6bf701e0eb592b24c3: This changes behavior: Previously there was a call to `error()`, now there is none?

maflcko reviewed (commented):
Happy to merge as-is. Let me know what you think.


review ACK ddbbdcf4112c3a84e3287e6bf701e0eb592b24c 🐯

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK ddbbdcf4112c3a84e3287e6bf701e0eb592b24c 🐯
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjssQwAplJ1ZbuRcMBriV0HxqVNf02T4stpYurghGghrlmhm8pidPOD4r+gcr+g
ECrmBrsbS9Ka/ycxwoM0nel0hEMsXLZWy6ueY0ufksVY8TFgIdVSQJtSLZ9MNuUE
7n9/Qz+z1tl2isShx5nbEQ+UcRiCfXVryfV0OuGO4RLPSJV9fqYaPBKh85af6hxt
P+q4qD4Ql6kWZLPyrdZ+M3JO3KIgRDdGLjXdvFQvNDVDos68HT+UTRvRod2B3kkM
5I84K4L6SfO/WW5sGXM/G8t7qyC4Sdv906qBP9zGkxFV/D5jj3n+cpgb8vog1RGr
6riVrC9pPnCC8kSopDgHdHIEiNdnqxeQJ2eiFutmI2B2Gz6ex7YjvBGyeJeRSka5
sNtrD+9EKtUE5Ta/Iby8mpl1+lSpzC9vFGJGuhTQg1kNzOV80MSVw480wDZcUrTM
vtZoxXpOH3/CnL+8Y+MvxanIahrvS/UApp60iZ848v8pEXtI4cGzUsysRIekUkeD
fmaWQU9d
=52YB
-----END PGP SIGNATURE-----
```


</details>

ryanofsky commented:
> Happy to merge as-is. Let me know what you think.

Thanks for review Marco. I definitely want to make some improvements based on your comments, so better not to merge for now

MarcoFalke commented on src/node/chain.cpp:
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2022 The Bitcoin Core developers
```
Maybe add this new file to iwyu?

MarcoFalke replied:
https://cirrus-ci.com/task/4504627910541312?logs=ci#L6998

MarcoFalke replied:
Looks like an iwyu bug, so maybe undo for now?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/25494#discussion_r921900387

> https://cirrus-ci.com/task/4504627910541312?logs=ci#L6998

> Looks like an iwyu bug, so maybe undo for now?

I'm not actually sure what the bug is but I added some missing includes and undid change that adds the file to IWYU for now. The larger PR https://github.com/bitcoin/bitcoin/pull/24230 this was part of adds more to this file so there is a chance to enable IWYU again later

ryanofsky reviewed (commented):
Updated ddbbdcf4112c3a84e3287e6bf701e0eb592b24c3 -> fb245ec3d3644cc9cef9b990ba2c8313277f5d41 ([`pr/ind.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.2) -> [`pr/ind.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ind.2..pr/ind.3)) with suggested changes like moving the new file from node/ to kernel/
Updated fb245ec3d3644cc9cef9b990ba2c8313277f5d41 -> f678151ff9a0a5934965b1c85ec97a1dfe45a67e ([`pr/ind.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.3) -> [`pr/ind.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ind.3..pr/ind.4)) to fix lint error https://cirrus-ci.com/task/6068993915092992

maflcko reviewed (commented):
f678151ff9a0a5934965b1c85ec97a1dfe45a67e 📭

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

f678151ff9a0a5934965b1c85ec97a1dfe45a67e 📭
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhL5AwAtnEh0cX8QFjSw7c20rH3WCFif2w8QDJpwnRDHHmc8qSzVamRRmmjGaZ2
uT2IvZCVzzLqvcJKWJR6Woan9pS5GJtF/pqBxv1olJd8CBNO1VrL3o5NnBzKIDAN
6PWq4Tl50TrIwnR5Wnd8pEYfbK77M4ghdSZ5TYtzL+gHTL8wQcYGmJCw7pDAgeAp
iMMe82rhI/sYeU9nqo3l5MF2IgJkzEuQ/wpxD4rmkDyf+9yLNtquHCBzR5BSPhDl
5LoR7v9LR5WqHZzOWyrrY/FhUmnbYDpmub8TAmYGj6cy+in37jLheDY4EDU2mzJG
CMtE/6yRGY8dPFohG//yUquEfHG7lVacW+9tMXrRZ0+zcmlBFcobOtwNRJ4Oc4qZ
IcmULASpFcVpFbLrOTqS5n/zrIvicQSxH4nXGXNnB7gvZFT4lmpnOin6txgN/Sic
sf20wlYEbQpJBWgBtRjSny8SuY5BHJeAypVzIbw1Sc78GpZs2m3bmjHZHuwud97p
FKbfDUd0
=d7Fu
-----END PGP SIGNATURE-----
```


</details>

ryanofsky reviewed (commented):
Updated f678151ff9a0a5934965b1c85ec97a1dfe45a67e -> 8bc1602821202d4863fab05ee1aab3ff3b70a7c0 ([`pr/ind.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.4) -> [`pr/ind.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ind.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ind.4..pr/ind.5)) updating commit message 8bc1602821202d4863fab05ee1aab3ff3b70a7c0 to describe some changed behavior and no longer enabling IWYU for now in the new file

maflcko reviewed (commented):
ACK 8bc1602821202d4863fab05ee1aab3ff3b70a7c0 though did not review the last commit 🗡

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 8bc1602821202d4863fab05ee1aab3ff3b70a7c0 though did not review the last commit 🗡
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhkpQv9EYKytsqsNKtA9Yte0NOoa/hTFzjqhx34b+kkeB9twTIhnl/8kqLcct6a
1983DDhe835/ZeKxwAGPV9KTnBma+E7Sr9nAaFuY8GJfA2qsFncXzoeRpZRW8e9m
EjZ1Yvi8Yac3AE9RArPayknO0uo7tvlYoL/2mx7k+6V/VvlZSAqrUMgwQY8PvQmJ
c6AoUvbUrMSy6HM4HZHi9TEVveQqwMqYwPwKRqy0A5CBAlwIN75m/NKFSDWcjVTS
p3m88uM5fZQ1gQ4Xu9FTvXSQavIKr883gUeUXOR9yv5FXgXHCRVGjrbpVEey2nmK
jOwdKWHVpVv0+6jI1qzJnkTlKdKvFKcvTdPp9EW/woWW4FdcEAtAG4s0kDfvQrN6
b8ZUUpowDJsM+3Uisuqq6dhq48j56u0Ux0RmXBTGsCT4t8hpPSUbsAqqSZnhpWlJ
SKw4c9amB3i8kcRLIBkpntEsOgtZnXBaNgdlpv3iWUALC8xqBOF8RDi10qMcS0x2
1/7likbu
=LR6p
-----END PGP SIGNATURE-----
```


</details>

ryanofsky commented:
@fjahr & @mzumsande could you reack? I think only change since previous reviews has been expanding the `m_block_index == nullptr` case in `CommitInternal` to keep logging an error and avoid committing anything

MarcoFalke commented:
ACK 7878f97bf15b6e7c9b47d1c0d96419b97e1bdcbd though did not review the last commit 🤼

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 7878f97bf15b6e7c9b47d1c0d96419b97e1bdcbd though did not review the last commit 🤼
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgZAAwAg2W4ylK74K+MHoP2XiZFNNZe6tzS3wpYnBUZXoP0ee1rOg/lNGl1o2ER
d0zs+tGjHajkpDiQB9DgCqNLQAEq1qOCRmxH1W9nt04gkVqyZPPcOg7GE2da1MNL
n8P8gNTHiqmS8FrRSAs40udE/VuhSr9j/4OLaCEjQiRdhWPKB1J4IduIK0OMQdW0
LQw2wBXRb0rtG4L0ztwmndFXx9IbBLRNZCFRR6a2IW4/2rZN3FmPSr6kDA7PV9S+
bM/+NNPThUga3KTfbPBj5ERivN/4UZXILyOdFVxs+dOcPViZVvtWd8uBLsiH+2QO
5zSuVeXnWPSujZXcNKpWnXbUJv7Pi8ClPkBv2016w5hzGGYOngV/TzLyeZtihn3S
IXQ7afP8zbCPBQvtcVUGB3QLZLvl1Byw88U+6OXSp0Yw/gEXXQLHujyLOI2G2/9P
RwNmi+6qDhhylB5TPist30uVk4D84NEDG/cz269cHe0+M+icnh2cpwysdM7fMO1S
vhWrJkyl
=8s9I
-----END PGP SIGNATURE-----
```


</details>

mzumsande commented:
Code Review ACK 7878f97bf15b6e7c9b47d1c0d96419b97e1bdcbd

ryanofsky commented on src/index/base.cpp:
```diff
@@ -359,7 +359,10 @@ bool BaseIndex::Start()
+    if (!Init()) return false;
+
+    const CBlockIndex* index = m_best_block_index.load();
+    if (!CustomInit(index ? std::make_optional(interfaces::BlockKey{index->GetBlockHash(), index->nHeight}) : std::nullopt)) {
```
In commit "indexes, refactor: Remove CBlockIndex* uses in index Init methods" (bef4e405f3de2718dfee279a9abff4daf016da26)

Calling the `CustomInit()` function after the `Init()` function here seems to cause a race condition, noticed by @furszy and described in https://github.com/bitcoin/bitcoin/pull/27607#issuecomment-1553400917. The problem is that `Init()` can set `m_synced` to true, which enables `BlockConnected` callbacks. But if `BlockConnected` callback runs before `CustomInit` finishes the index may still be only partially loaded, and the `CustomAppend` call in `BlockConnected` could fail
